### PR TITLE
CompositeKey validation checks

### PR DESCRIPTION
**Pull request by kchalkias**
Some extra validation checks for CompositeKeys
1) Min threshold = 1
2) Min weight = 1
3) threshold <= totalWeight (total = aggregated)
4) Graph cycle detection (to avoid infinite loops)
5) totalWeight integer overflow detection (inherently,  due to 1 & 2, we protect against underflow as well)

Also, because `@CordaSerializable` doesn't invoke the constructor and thus a key may be malformed (i.e. cannot be validated), `isFulfilledBy()` runs a full validity check before checking thresholds are satisfied.

There is also a small fix on the "N of N" requirement, where the total (aggregated) weight value is now taken into account, if threshold is not provided. Previously this was `children.size`.

~_This is not subject for review yet; unit-tests are required._~ Ready for review.

### CODE REVIEW

**Comment by mikehearn** [comment]
> Also, because @CordaSerializable doesn't invoke the constructor

That's fixed as part of Rick's serialisation work I think, @rick-r3 ?

**Comment by mikehearn** [comment]
BTW to detect overflow just use Math.addExact and let it throw.

**Comment by kchalkias** [comment]
@mikehearn I had a look on the `Math.addExact` already, but I ended up on this solution because
a) I do an extra internal test, for negative weight (that is application specific)
b) is `addExact` faster? cause it involved more operations as it works generically for both negative and positive numbers
```
public static int addExact(int x, int y) {
        int r = x + y;
        // HD 2-12 Overflow iff both arguments have the opposite sign of the result
        if (((x ^ r) & (y ^ r)) < 0) {
            throw new ArithmeticException("integer overflow");
        }
        return r;
}
```

**Comment by kchalkias** [comment]
@mikehearn, there is a TODO to remove the extra validity test on `isFullfilledBy`, this is temporary

**Comment by mikehearn** [comment]
Math.addExact compiles down to a single `add/sub; jo` sequence where the jo points to a deopt handler, so it's about as efficient as you can get.

**Comment by kchalkias** [comment]
Ok thanks, current approach just checks for `sum>0` which is also a single operation plus we have a more detailed error description, but maybe `addExact` is more elegant :)

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 1]
```diff
@@ -28,16 +29,81 @@ class CompositeKey private constructor (val threshold: Int,
...
+     * traversing the graph and key duplicates in the each layer. It also checks if the threshold and weight constraint
+     * requirements are met, while it tests for aggregated-weight integer overflow.
+     * In practice, this method should be always invoked on the root [CompositeKey], as it inherently
+     * validates the child nodes (all the way till the leaves).
+     * TODO: Always call this method when deserialising [CompositeKey]s.
+     */
+    fun checkValidity() {
+        val visitedMap = IdentityHashMap<CompositeKey,Boolean>()
+        visitedMap.put(this, true)
+        cycleDetection(visitedMap) // Graph cycle testing on the root node.
```
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: Would it be worth inlining cycle detection into `checkValidity` so the graph is only traversed once, or does it make this function too messy?

File: core/src/test/kotlin/net/corda/core/crypto/CompositeKeyTests.kt [hunk 2]
```diff
@@ -124,4 +124,139 @@ class CompositeKeyTests {
...
+        assertFailsWith(IllegalArgumentException::class) {
+            val node1 = CompositeKey.Builder().addKeys(alicePublicKey, bobPublicKey).build()
+            val node2 = CompositeKey.Builder().addKeys(bobPublicKey, alicePublicKey).build()
+            CompositeKey.Builder().addKeys(node1, node2).build()
+        }
+    }
+
+    @Test()
+    fun `composite key validation with graph cycle detection`() {
+        val node1 = CompositeKey.Builder().addKeys(alicePublicKey, bobPublicKey).build() as CompositeKey
```
Comment (rnicoll) [inline core/src/test/kotlin/net/corda/core/crypto/CompositeKeyTests.kt]: Would `key1`, `key2`, etc. perhaps be clearer? `node1` sounds like a Corda node rather than graph node, as that's how we normally use the term

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 3]
```diff
@@ -75,21 +145,27 @@ class CompositeKey private constructor (val threshold: Int,
...
+        return combinedWeight >= threshold
+    }
+
+    /**
+     * Function checks if the public keys corresponding to the signatures are matched against the leaves of the composite
+     * key tree in question, and the total combined weight of all children is calculated for every intermediary node.
+     * If all thresholds are satisfied, the composite key requirement is considered to be met.
+     */
+    fun isFulfilledBy(keysToCheck: Iterable<PublicKey>): Boolean {
+        checkValidity() // TODO: remove when checkValidity() will be eventually invoked during deserialization.
```
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: Wouldn't it make more sense of `checkValidity()` to go into either the builder or an `init { }` block directly? It feels like we're making changes we'll just have to roll back later.

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 4]
```diff
@@ -75,21 +145,27 @@ class CompositeKey private constructor (val threshold: Int,
...
-    /**
-     * Function checks if the public keys corresponding to the signatures are matched against the leaves of the composite
-     * key tree in question, and the total combined weight of all children is calculated for every intermediary node.
-     * If all thresholds are satisfied, the composite key requirement is considered to be met.
-     */
-    fun isFulfilledBy(keysToCheck: Iterable<PublicKey>): Boolean {
+    // Extracted method from isFulfilledBy.
+    private fun checkFulfilledBy(keysToCheck: Iterable<PublicKey>): Boolean {
         if (keysToCheck.any { it is CompositeKey } ) return false
-        val totalWeight = children.map { (node, weight) ->
```
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: Can we rename this back to `totalWeight` unless there's a good reason for the change, that I'm missing?

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt [hunk 5]
```diff
@@ -28,16 +29,81 @@ class CompositeKey private constructor (val threshold: Int,
...
+     * In practice, this method should be always invoked on the root [CompositeKey], as it inherently
+     * validates the child nodes (all the way till the leaves).
+     * TODO: Always call this method when deserialising [CompositeKey]s.
+     */
+    fun checkValidity() {
+        val visitedMap = IdentityHashMap<CompositeKey,Boolean>()
+        visitedMap.put(this, true)
+        cycleDetection(visitedMap) // Graph cycle testing on the root node.
+        checkConstraints()
+        for ((node) in children) {
```
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: Does `(node)` actually need brackets?

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt (hunk 3 above)
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: The problem is when you deserialise a potentially "altered" CompositeKey. I didn't add it to the Builder or init, cause the cycle check is practically required on the root node only. It doesn't harm to invoke it on the creation of any sub-node object, but it may introduce a performance penalty.

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt (hunk 4 above)
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: In my mind, totalWeight should include each party (total weight for this Composite Key - before signatures). When we collect signatures it is actually the collected-signature-total or currently aggregated; just to distinguish between full total and current total

File: core/src/test/kotlin/net/corda/core/crypto/CompositeKeyTests.kt (hunk 2 above)
Comment (kchalkias) [inline core/src/test/kotlin/net/corda/core/crypto/CompositeKeyTests.kt]: I used the `tree canonical form` as a basis for these tests, but you are right, I can change it.

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt (hunk 3 above)
Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: Ah... in that case I think change the comment here, to something like:

`// We validate keys only when checking if they're matched, as this checks subkeys as a result. Doing these checks at deserialization/construction time would result in duplicate checks`

Also perhaps add a transient `var` that indicates if a key has been validated successfully, so these checks are only run at most once per instance?

File: core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt (hunk 5 above)
Comment (kchalkias) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: @rnicoll it's a de-structuring declaration omitting weight. If I remove it then it is of NodeAndWeight type.
Do you want to change it to `(node, _)` for clarity?

Comment (rnicoll) [inline core/src/main/kotlin/net/corda/core/crypto/CompositeKey.kt]: Ah, yes, that would be clearer, thanks!

**Comment by kchalkias** [comment]
@rnicoll thanks for the review, using the `transient var validated` resolved the inline `cycleDetection` requirement as well.
